spark/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableStrategy.scala

```diff
@@ -0,0 +1,267 @@
+        sparkSession, ClusteringColumnInfo.extractLogicalNames(snapshot))
+    case OptimizeTableMode.ZORDER => ZOrderOptimizeTableStrategy(sparkSession, zOrderBy)
+    case OptimizeTableMode.COMPACTION =>
+      CompactionOptimizeTableStrategy(sparkSession, optimizeContext)
+    case other => throw new UnsupportedOperationException(s"Unsupported mode $other")
+  }
+
+  private def getMode(snapshot: Snapshot, zOrderBy: Seq[String]): OptimizeTableMode.Value = {
+    val isClusteredTable = ClusteredTableUtils.isSupported(snapshot.protocol)
+    val isZOrderBy = zOrderBy.nonEmpty
+    if (isClusteredTable) {
+      assert(!isZOrderBy)
+      OptimizeTableMode.CLUSTERING
+    } else if (isZOrderBy) {
+      OptimizeTableMode.ZORDER
+    } else {
+      OptimizeTableMode.COMPACTION
+    }
+  }
+}
+
+/** Implements compaction strategy */
+case class CompactionOptimizeTableStrategy(
+    override val sparkSession: SparkSession,
+    optimizeContext: DeltaOptimizeContext) extends OptimizeTableStrategy {
+
+  override val optimizeTableMode: OptimizeTableMode.Value = OptimizeTableMode.COMPACTION
+
+  // In COMPACTION, all files within a bin are written into single larger file.
+  override val maxBinSize: Long = {
+    optimizeContext.maxFileSize.getOrElse(
+      sparkSession.sessionState.conf.getConf(DeltaSQLConf.DELTA_OPTIMIZE_MAX_FILE_SIZE))
+  }
+
+  override def curve: String = {
+    throw new IllegalArgumentException("Compaction doesn't support clustering.")
+  }
+
+  override def updateOptimizeStats(
+      optimizeStats: OptimizeStats,
+      removedFiles: Seq[RemoveFile],
+      bins: Seq[(Map[String, String], Seq[AddFile])]): Unit = {}
+}
+
+/** Implements ZOrder strategy */
+case class ZOrderOptimizeTableStrategy(
+    override val sparkSession: SparkSession,
+    zOrderColumns: Seq[String]) extends OptimizeTableStrategy {
+
+  assert(zOrderColumns.nonEmpty)
+
+  override val optimizeTableMode: OptimizeTableMode.Value = OptimizeTableMode.ZORDER
+
+  override val curve: String = "zorder"
+
+  // For ZORDER, set maxBinSize the maximal LONG value to have single BIN for each partition.
+  override val maxBinSize: Long = Long.MaxValue
+
+  override def updateOptimizeStats(
+      optimizeStats: OptimizeStats,
+      removedFiles: Seq[RemoveFile],
+      bins: Seq[(Map[String, String], Seq[AddFile])]): Unit = {
+    val inputFileStats =
+      ZOrderFileStats(removedFiles.size, removedFiles.map(_.size.getOrElse(0L)).sum)
+    optimizeStats.zOrderStats = Some(ZOrderStats(
+      strategyName = "all", // means process all files in a partition
+      inputCubeFiles = ZOrderFileStats(0, 0),
+      inputOtherFiles = inputFileStats,
+      inputNumCubes = 0,
+      mergedFiles = inputFileStats,
+      // There will one z-cube for each partition
+      numOutputCubes = optimizeStats.numPartitionsOptimized))
+  }
+}
+
+/** Implements clustering strategy for clustered tables */
+case class ClusteringOptimizeTableStrategy(
```

Comment:
Can we simply do `CompactionStrategy`, `ZOrderStrategy` and `ClusteringStrategy`?